recipes/AISHELL-1/ASR/transformer/train.py

```diff
@@ -122,37 +122,17 @@ def compute_objectives(self, predictions, batch, stage):
             self.acc_metric.append(p_seq, tokens_eos, tokens_eos_lens)
         return loss
 
-    def fit_batch(self, batch):
-        """Train the parameters given a single batch in input"""
+    def on_fit_batch_start(self, batch, should_step):
+        """Gets called at the beginning of each fit_batch."""
         # check if we need to switch optimizer
         # if so change the optimizer from Adam to SGD
         self.check_and_reset_optimizer()
 
-        predictions = self.compute_forward(batch, sb.Stage.TRAIN)
-        loss = self.compute_objectives(predictions, batch, sb.Stage.TRAIN)
-
-        # normalize the loss by gradient_accumulation step
-        (loss / self.hparams.gradient_accumulation).backward()
-
-        if self.step % self.hparams.gradient_accumulation == 0:
-            # gradient clipping & early stop if loss is not fini
-            self.check_gradients(loss)
-
-            self.optimizer.step()
-            self.optimizer.zero_grad()
-
-            # anneal lr every update
+    def on_fit_batch_end(self, batch, outputs, loss, should_step):
+        """At the end of the optimizer step, apply noam annealing."""
+        if should_step:
             self.hparams.noam_annealing(self.optimizer)
```

Comment:
When the annealing happens at the end of `fit_batch`, doesn't the very first batch use a non-annealed LR?

I think I ran into some problems with this in my own recipes, where the first step uses a large LR and takes way too big of a step (gradients can be very large at first step too).

This connected to the issue of SpeechBrain not using the PyTorch scheduler interface.

Reply:
Not that this applies to the other recipes as well, I just used this as an example, so if I am right, this could be an important thing to fix.

Reply:
Hmmmm how would you deal with it? Doing it at the beginning of the `fit_batch` ? Also, the LR annealing is only applied every `should_step`.

Reply:
OK. Sorry I didn't read at first your big comment :p. It could be a way. Using LR annealing at the beginning of the fit_batch fn. I propose with this PR to solely target the mentioned issue (i.e., DDP, precision, overriding). We could think about adding a better interface with PyTorch scheduling in another PR. But I tend to fully agree with you.

Reply:
Yeah, I think the simple fix would be to update the LR at the beginning of fit_batch ( conditional on should_step of course, but before the first optimizer step).